docs_src/design/adr/security/0009-Secure-Bootstrapping.md

```diff
@@ -0,0 +1,463 @@
+Vault access tokens have a TTL
+and it is not possible to renew an expired token.
+New ones must be generated whenever the framework is started.
+
+Though Vault access tokens cannot be bootstrapped at install time,
+it is nevertheless important to hold off starting
+the majority of EdgeX services
+until Vault is up and access tokens generated.
+
+#### Consul access tokens
+
+Consul access tokens can be requested from the Vault Consul secrets engine at runtime
+on an as-needed basis.
+
+#### TLS assets for EdgeX and user microservices
+
+These TLS assets must be generated at runtime.
+For efficiency, they can be generated upon upon first use
+or near the expiration window of current certificates,
+and cached in the service's secret store.
+
+### Solutions ruled out
+
+#### Manual secret provisioning
+
+A typical cloud-based microservice architecture typically has a manual provisioning step.
+This step would include activities such as configuring Vault,
+installing a database schema,
+setting up database service account passwords,
+and seeding initial secrets such as PKI private keys
+that have been generated offline
+(possibly requiring several days of lead time).
+A cloud team may have weeks or months to prepare for this event,
+and it might take the greater part of a day.
+
+In contrast,
+EdgeX up to this point has been a "turnkey" middleware framework:
+it can be deployed with the same ease as an application,
+such as via a docker-compose file,
+or via a snap install.
+This means that most of the secret provisioning must be automated
+and the provisioning logic must be built into the framework in some way.
+
+For the above reason,
+manual secret provisioning should not be considered as a solution for secure bootstrapping.
+
+## Decision
+
+### Stage-gate mechanism
+
+#### Stage-gate requirements
+
+The stage-gate mechanism must control the bootstrapping flow
+in the OCI container deployment scenario,
+where the underlying container orchestration
+does not provide a supported "install" flow.
+The docker-compose tool is one such framework.
+
+The following stages are defined:
+
+* **Install stage**: Used for a fresh EdgeX installation.
+* **Upgrade stage**: Used to upgrade from one EdgeX release to another.
+* **Run stage**: The stage where EdgeX services are running in production.
+
+In particular, it is necessary to forestall entry into the "run stage"
+until either the install stage or the upgrade stage has completed.
+There is currently no "uninstall stage" defined.
+
+The stage-gate mechanism must work in the following environments:
+
+* docker-compose in Linux on a single node
+* docker-compose in Microsoft* Windows* on a single node
+* docker-compose in Apple* MacOS* on a single node
+
+#### Stage-gate design
+
+Docker-compose v2.x used to have a `depends_on / condition` directive
+that would test a service's HEALTHCHECK
+and block startup
+until the service was "healthy".
+Unfortunately, this feature was removed in 3.x docker-compose.
```

Comment:
a note: `depends on` also ignored in docker-swarm mode

Reply:
added